scanner/scanner.go

```diff
@@ -14,7 +13,7 @@ import (
 // ScannerAble is implemented by zbuf.Readers that provide an optimized
 // implementation of the Scanner interface.
 type ScannerAble interface {
-	NewScanner(ctx context.Context, filterExpr ast.BooleanExpr, s nano.Span) (Scanner, error)
+	NewScanner(ctx context.Context, filter filter.Filter, span nano.Span) Scanner
```

Comment:
I'm planning to extract the Boyer-Moore search pattern from this ast.BooleanExpr in zngio.Reader.NewScanner, so I think this needs to stay as it was.

Reply:
In e47a3366a3e5078c82e9ee1cae89f20e038717d8, I adjusted this to take both the compiled filter & the filter expression. I already had both in the `driver.SearchFilter` struct, as we'll eventually want to use the filter expression to prune data in the `SendSources` call. Let me know what you think.

Reply:
I want to keep the second return parameter so I can return an error after examining the ast.BooleanExpr.

And given that, I think the interface is nicer with just the ast.BooleanExpr. For example, with just that you can't accidentally pass mismatched parameters.

Reply:
How about if we create a new struct that has both the filter expression & the filter, like:
```go
struct CompiledFilter {
Expr ast.BooleanExp
Matcher filter.Filter
}
```

and then `NewScanner` would take that? I'd like to avoid needing to re-compile the same filter expression for every file in an archive.

Reply:
I prefer separate arguments to a new structure.

scanner.ScannerAble.NewScanner and scanner.NewScanner don't need to have the same signature, and the latter's the one that matters for this PR, so how about leaving scanner.ScannerAble.NewScanner unchanged and changing scanner.NewScanner to this (with a second return parameter for an error)?
```go
func NewScanner(ctx context.Context, r zbuf.Reader, filter filter.Filter, filterExpr ast.BooleanExpr, s nano.Span) (Scanner, error) {
```

Reply:
Oh, idea that doesn't make any sense.

Reply:
My first preference is to leave scanner.ScannerAble.NewScanner and scanner.NewScanner as they were. The interfaces are simpler, compiling the filter doesn't strike me as too expensive to do once per file, and if proves wrong we can always add the filter.Filter parameter later.

My second preference is to change both to have this signature.
```go
NewScanner(ctx context.Context, r zbuf.Reader, f filter.Filter, filterExpr ast.BooleanExpr, s nano.Span) (Scanner, error)
```

Reply:
added the new signature in 17f84c5